### ISSUE TITLE
Distance Progress Logic for Flipped design

### ISSUE THREAD

**Issue by terracoda** [phetsims/ratio-and-proportion#262]
While working on lock ratio responses, and thinking about a design flip in https://github.com/phetsims/ratio-and-proportion/issues/245, I realized new logic for providing distance progress information would be beneficial for all perspectives of the moving the hands, either individually or at the same time.

This logic is demonstrated in examples in the design document in these tables: 
- [Table: Test of design flip with new qualitative distance regions - Moving Left Hand](https://docs.google.com/document/d/1HJvvGWYaUv9tlat9S7omQQVng8fGB4FuUhTEM1qx1z8/edit?ts=5fc56c26#heading=h.hq2zvuiwvq6f)
- [Table: LOCK Ratio - Starting near bottom and moving up to top](https://docs.google.com/document/d/1HJvvGWYaUv9tlat9S7omQQVng8fGB4FuUhTEM1qx1z8/edit?ts=5fc56c26#heading=h.8nbwkxiwy8lf)

There are other examples, too, but they may not be using the latest scales for distance and Play Area position.

The summary of the new logic can be found [here in the design doc](https://docs.google.com/document/d/1HJvvGWYaUv9tlat9S7omQQVng8fGB4FuUhTEM1qx1z8/edit?ts=5fc56c26#heading=h.x0puf5gck72k) and here is copied for convenience:

Once we flip the design in issue #245, context responses will just have qualitative distance regions (QDR) and play area position region (PAR).

1. Never replace a context response if it contains a play area region (PAR) landmark, that is, if it contains a single-value pa-region (“at top”, “at bottom”, “at middle”). I see these single-value regions as important landmarks, and would like them to be persistent.
2. Never replace a context response if it contains a distance region (QDR) landmark, that is, if it contains “farthest from” or “even with”. Again, I see these as important distance markers that should be persistent.
3. Never replace the context response on the last step that takes you to the min or min position (this may already be covered by number 1).
4. Otherwise, replace entire context response (qual-distance-region & pa-region) with a distance progress response (i.e., {{farther apart}}, {{closer together}}) the first time the QDR repeats (.e.g, two consecutive steps in the same qual-distance-region (QRD)).

**Comment by terracoda** [comment]
This new logic may not work as well as the current logic does, when Tick Marks are checked.
The logic can quote easily start a pattern that replaces every or almost every even-numbered Tick Mark. 
This is undesirable.

**Comment by terracoda** [comment]
Ok, I tested out some more scenarios and would like to discuss additions to my proposed logic to make things better with the tick marks on. Here's the new logic that seems logical to me:
**As above:**
1. Never replace a context response if it contains a play area region landmark, that is, if it contains a single-value pa-region (“at top”, “at bottom”, “at middle”). I see these single-value regions as important landmarks, and would like them to be persistent.
2. Never replace a context response if it contains a distance region landmark, that is, if it contains “farthest from” or “even with”. Again, I see these as important distance markers that should be persistent.
3. Never replace the context response on the last step that takes you to the min or min position (this may already be covered by number 1).
4. Otherwise, replace entire context response (qual-distance-region & pa-region) with a distance progress response (i.e., {{farther apart}}, {{closer together}}) the first time the qual-distance-region  repeats (.e.g, two consecutive steps in the same qual-distance-region).
**AND in addition to the above above:**
5. If Lock Ratio is NOT checked, never replace the context response when the hands are moved into the challenge ratio, similar to item 3.
6. When using the sliders, when replacement happens on an even tick mark, keep the tick mark value in the response.

I have more examples added to the Table: Testing design flip in the design doc.

**Comment by zepumph** [comment]
Above is quite complicated. I am happy to continue down that road, but as a first-pass, I decided to approach this in the opposite manner. Instead out outlining each edge case that we wouldn't want to specify an overriding direction description, I just chose a simple criteria that compares the whole alert instead of just the QDR. I'm not sure how far this will take us, but I'm excited to be able to use it to progress the conversation we will do tomorrow. Commit coming soon.

**Comment by zepumph** [comment]
After a discussion with @terracoda, we decided to make a minor change to my commit/her design where instead swapping the entire context response, we will only ever swap out the QDR when it repeats. This way we will always have the position in the context response. 

- [ ] There is a chance that this will still be too long, especially for both hands, but we will need to implement first before deciding with we need shortening.

- [ ] We will need to add in (5) (from https://github.com/phetsims/ratio-and-proportion/issues/262#issuecomment-738125761) once an initial commit comes in, since we could have "closer to" or "farther from" when "at ratio" right now. It may be enough to just provide position, but I will write a not here to revisit.

**Comment by zepumph** [comment]
Notes: 
- [ ] Comma after "Hands. . . " for distance and proximity to ratio?

**Comment by zepumph** [comment]
Everything I can think of has been implemented above.  The checkboxes that aren't checked in https://github.com/phetsims/ratio-and-proportion/issues/262#issuecomment-739020593 need to have a bit of discussion, and likely for @terracoda to review before we discuss next Tuesday. I'm going to mark this for meeting, because there is a lot to discuss, but we can still make progress on it in the interim.

**Comment by zepumph** [comment]
> Comma after "Hands. . . " for distance and proximity to ratio?


@terracoda said that it is correct as is. The object response gets a comma, like `hands, extremely close to challenge ratio`, but the context response (capitalized) doesn't get one, "Hands very close to each other, left on tick mark 2, right on tick mark 4.5."


>What do we want to test against for "duplication", since single/both use the same type of alert, it is possible to set it up so that if you get a distance from the both-hands, then tab to an individual and get the same distance, that would count as a repeat. For now I'll implement both separately.

We are ok with one underlying variable that runs the duplication-testing logic for all input modalities (both hands and individual sliders).

>Right now we are counting closer/farther as a previous so that we don't ever get two of them at the same time. Another way we could implement would be to only keep track of the QDR, and so we would get the region when it changes, and then each subsequent time before the new region, we would get closer/farther info. This is more like what was described in #220 (comment). I don't think we want to do this, but it is worth discussing.

We like it the way it alternates back and forth now.

>This logic is based on when alerts get sent to the utteranceQueue, not if they are actually heard. It is easy to press twice quickly and have the first alert be "farther from".

@terracoda and I feel like this is acceptable, especially since there is nothing that we can do. The swap also doesn't trade out "singleton" information, as there is always position, and it is unlikely that a user could interact in enough even-only numbers of keypresses for this to be a problem

**Comment by zepumph** [comment]
>We are ok with one underlying variable that runs the duplication-testing logic for all input modalities (both hands and individual sliders).

After making it one variable, it actually had no change in logic, since the both-hands used a lowercase version of the QDR, and individual sliders were capitalized. I don't really think it matters either way, so I made the code that much simpler.

**Comment by terracoda** [comment]
@zepumph, the labels are still meeting and ready for review. Please let me know when this issue, #245 and #261 are all ready for review :-)

**Comment by zepumph** [comment]
Ready.

**Comment by terracoda** [comment]
Ok, the adjusted logic we discussed is working.
I think we may not need to make additional changes outlined in https://github.com/phetsims/ratio-and-proportion/issues/262#issuecomment-738958504

I think, to be sure we need to revisit this issue after #279 and #209.

I find when the position of each hand includes "around half-way past" or "around X.5 Tick Mark" things are too long.

Having the much simpler qual-distance-region description or the shorter distance-progress next to those longer position strings, is certainly helpful.

Hopefully, once we address #279 AND #209, we might not hear those longer position responses as often as we do now.

**Comment by terracoda** [comment]
I'll mark this one verified and on-hold. I can listen again after #209 and #279 are closed.

**Comment by terracoda** [comment]
Added the meeting label, that should help us not forget about this issue.

**Comment by zepumph** [comment]
@terracoda, I wasn't sure what you wanted to get out of this issue from the meeting today. Do you want to talk more about this? I'm happy to chat if we want this before next week.

**Comment by terracoda** [comment]
I am doing another review now that the above mentioned issues are addressed.

**Comment by zepumph** [comment]
During a meeting with @terracoda, we decided that the checkboxes in https://github.com/phetsims/ratio-and-proportion/issues/262#issuecomment-738958504 don't need to be done. I added strikethrough to them.

**Comment by terracoda** [comment]
Agreed. Everything in this issue has been addressed.
Closing this issue.

**Comment by terracoda** [comment]
I'll add that the one thing we lost with the design flip was a mention of the position of each hand directly in the PDOM nearby the interactions. The object value of the sliders used to contain hand position. It now contains "proximity to ratio", and position is in context responses.

To address this minor gap, I'll open new issue to flesh out a short dynamic description of the hands.
